server/routes/index.ts

```diff
@@ -745,20 +745,39 @@ export function defineRoutes(router: IRouter) {
   router.delete(
     {
       path: `${API_PREFIX}/configuration/cache`,
-      validate: false,
+      validate: {
+        body: schema.object({
+          dataSourceId: schema.maybe(schema.string()),
+        }),
+      },
     },
     async (context, request, response) => {
-      const client = context.security_plugin.esClient.asScoped(request);
-      let esResponse;
-      try {
-        esResponse = await client.callAsCurrentUser('opensearch_security.clearCache');
-        return response.ok({
-          body: {
-            message: esResponse.message,
-          },
-        });
-      } catch (error) {
-        return errorResponse(response, error);
+      if (!dataSourceEnabled || !request.body?.dataSourceId) {
+        const client = context.security_plugin.esClient.asScoped(request);
+        let esResponse;
+        try {
+          esResponse = await client.callAsCurrentUser('opensearch_security.clearCache');
+          return response.ok({
+            body: {
+              message: esResponse.message,
+            },
+          });
+        } catch (error) {
+          return errorResponse(response, error);
+        }
+      } else {
+        const client = context.dataSource.opensearch.legacy.getClient(request.body?.dataSourceId);
+        let esResponse;
+        try {
+          esResponse = await client.callAPI('opensearch_security.clearCache', {});
+          return response.ok({
+            body: {
+              message: esResponse.message,
+            },
+          });
+        } catch (error) {
+          return errorResponse(response, error);
+        }
```

Comment:
can we extract this block as a separate function since its duplicate of L757-L767?

Reply:
Done